[webkitapipy] Support conditional allowlist entries

emw-apple opened the pull request:
#### 2bb7daae7de66f47266282387a2a6ff60ee62b25
<pre>
[webkitapipy] Support conditional allowlist entries
<a href="https://bugs.webkit.org/show_bug.cgi?id=295681">https://bugs.webkit.org/show_bug.cgi?id=295681</a>

Reviewed by Sam Sneddon.

Some allowed SPI are only used in specific build configurations. Since
audit-spi checks allowlists for exhaustiveness, we need a way to only
allow SPI based on specific build-time conditions.

Building off of the &lt;wtf/Platform.h&gt; flags generator added earlier this
year, teach audit-spi to ingest `-D` compilation condition flags, the
same way swift and clang do. Have it read the response file of platform
flags the same way that swift does.

Support an optional &quot;requires&quot; list in allowlist entries. For example,
the following SPI is only called in WebAuthn code, so we could write:

    [passkeys.equivalent-api]
    symbols = [&quot;_kSecAttrAlias&quot;]
    requires = [&quot;ENABLE_WEB_AUTHN&quot;]

If we supported building for iOS without ENABLE_WEB_AUTHN, this
`requires` list would make it possible to SPI-check both configurations.

Requirements can be a conjunction of multiple conditions. A requirement
can check for the absence of a condition by prefixing it with &quot;!&quot;. For
example:

    requires = [&quot;ENABLE_FOO&quot;, &quot;!ENABLE_BAR&quot;]

This matches under the same conditions that
`ENABLE(FOO) &amp;&amp; !ENABLE(BAR)` would in the C preprocessor.

--

Since allowlist matching is done inside the SDKDB cache, the
implementation works by expressing conditions to the sqlite database. A
`condition_chain` table represents requirements lists as boolean
variables in a graph data structure. Like other aspects of allowlists,
these are persisted and only update when an allowlist changes.

At runtime, audit-spi processes `-D` arguments and fills in an ephemeral
table of active conditions. In SDKDB.audit(), it traverses all edges in
condition_chain to find active &quot;chains&quot; of variables, where every
variable in the chain is satisfied. An allowlist entry is only matched
if it is associated with an active chain, or has no chain.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj: Run
  `generate-platform-args` in JSC&apos;s derived sources target, just like
  other projects do. This is the first time it&apos;s used from JSC, since
  JSC has no swift code.
* Source/WTF/Scripts/audit-spi-if-needed.sh: Pass in the generated
  response file to audit-spi.
* Source/WTF/Scripts/generate-platform-args: Tweak to work in JSC&apos;s
  build, where some build settings are undefined because they are not
  customized.
* Source/WebKit/Configurations/AllowedSPI-legacy.toml: There is one
  allowed class name that changes based on whether we are building for
  internal or open source. Use requirements to catch it.
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj: Run
  `generate-platform-args`, same as JSC change.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
(AllowedSPI):
(AllowList.from_dict):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow_unittest.py:
(TestAllowList.test_no_repetition_requires):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
(get_parser): Support reading arguments from compiler-style response
  files.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
(SDKDB._initialize_db):
(SDKDB._initialize_temporary_schema):
(SDKDB.InsertionKind):
(SDKDB.InsertionKind.statement):
(SDKDB._add_allowlist):
(SDKDB):
(SDKDB.add_defines):
(SDKDB.audit):
(SDKDB._add_objc_interface):
(SDKDB._add_symbol):
(SDKDB._add_objc_class):
(SDKDB._add_objc_selector):
(SDKDB.InsertionKind.table_name): Deleted.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_unittest.py:
(TestSDKDB.test_audit_allowed_conditional):
(TestSDKDB):
(TestSDKDB.test_audit_missing_name_conditional):
(TestSDKDB.test_audit_missing_name_negated_conditional):
(TestSDKDB.test_audit_allowed_negated_conditional):
(TestSDKDB.test_audit_allowed_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions_negation):

Canonical link: <a href="https://commits.webkit.org/297295@main">https://commits.webkit.org/297295@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/73814b5424152106d2c1ecf1ff831bcaf4e31075

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111216 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30882 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21328 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117247 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/61484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113178 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31563 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39464 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84535 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/61484 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114163 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25209 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100127 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64981 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/110649 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24551 "Passed tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/135/builds/18268 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61067 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/103707 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94587 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18335 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120315 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/109769 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38265 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28433 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/93462 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38641 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96402 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93287 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23774 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38382 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/16145 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/34250 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38154 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43631 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/134045 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37819 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/36149 "Found 2 new JSC stress test failures: wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention-2.js.default, wasm.yaml/wasm/stress/jitless-wasm-to-js-should-follow-to-wasm-call-frame-convention.js.default (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41152 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39521 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

gsnedders commented on Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
```diff
@@ -316,15 +320,39 @@ def add_allowlist(self, allowlist: Path) -> bool:
+                # cross-references condition chains with active conditions
+                # added by add_defines().
+                #
+                # FIXME: No effort is made to reuse nodes in the graph between
+                # allowlist entries, so we store more than we have to.
```
Can you file a bug to track this and link it?

emw-apple replied:
Yes, thanks! https://bugs.webkit.org/show_bug.cgi?id=295819

emw-apple committed:
[webkitapipy] Support conditional allowlist entries
https://bugs.webkit.org/show_bug.cgi?id=295681

Reviewed by Sam Sneddon.

Some allowed SPI are only used in specific build configurations. Since
audit-spi checks allowlists for exhaustiveness, we need a way to only
allow SPI based on specific build-time conditions.

Building off of the <wtf/Platform.h> flags generator added earlier this
year, teach audit-spi to ingest `-D` compilation condition flags, the
same way swift and clang do. Have it read the response file of platform
flags the same way that swift does.

Support an optional "requires" list in allowlist entries. For example,
the following SPI is only called in WebAuthn code, so we could write:

    [passkeys.equivalent-api]
    symbols = ["_kSecAttrAlias"]
    requires = ["ENABLE_WEB_AUTHN"]

If we supported building for iOS without ENABLE_WEB_AUTHN, this
`requires` list would make it possible to SPI-check both configurations.

Requirements can be a conjunction of multiple conditions. A requirement
can check for the absence of a condition by prefixing it with "!". For
example:

    requires = ["ENABLE_FOO", "!ENABLE_BAR"]

This matches under the same conditions that
`ENABLE(FOO) && !ENABLE(BAR)` would in the C preprocessor.

--

Since allowlist matching is done inside the SDKDB cache, the
implementation works by expressing conditions to the sqlite database. A
`condition_chain` table represents requirements lists as boolean
variables in a graph data structure. Like other aspects of allowlists,
these are persisted and only update when an allowlist changes.

At runtime, audit-spi processes `-D` arguments and fills in an ephemeral
table of active conditions. In SDKDB.audit(), it traverses all edges in
condition_chain to find active "chains" of variables, where every
variable in the chain is satisfied. An allowlist entry is only matched
if it is associated with an active chain, or has no chain.

* Source/JavaScriptCore/JavaScriptCore.xcodeproj/project.pbxproj: Run
  `generate-platform-args` in JSC's derived sources target, just like
  other projects do. This is the first time it's used from JSC, since
  JSC has no swift code.
* Source/WTF/Scripts/audit-spi-if-needed.sh: Pass in the generated
  response file to audit-spi.
* Source/WTF/Scripts/generate-platform-args: Tweak to work in JSC's
  build, where some build settings are undefined because they are not
  customized.
* Source/WebKit/Configurations/AllowedSPI-legacy.toml: There is one
  allowed class name that changes based on whether we are building for
  internal or open source. Use requirements to catch it.
* Source/WebKitLegacy/WebKitLegacy.xcodeproj/project.pbxproj: Run
  `generate-platform-args`, same as JSC change.

* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow.py:
(AllowedSPI):
(AllowList.from_dict):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/allow_unittest.py:
(TestAllowList.test_no_repetition_requires):
* Tools/Scripts/libraries/webkitapipy/webkitapipy/program.py:
(get_parser): Support reading arguments from compiler-style response
  files.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb.py:
(SDKDB._initialize_db):
(SDKDB._initialize_temporary_schema):
(SDKDB.InsertionKind):
(SDKDB.InsertionKind.statement):
(SDKDB._add_allowlist):
(SDKDB):
(SDKDB.add_defines):
(SDKDB.audit):
(SDKDB._add_objc_interface):
(SDKDB._add_symbol):
(SDKDB._add_objc_class):
(SDKDB._add_objc_selector):
(SDKDB.InsertionKind.table_name): Deleted.
* Tools/Scripts/libraries/webkitapipy/webkitapipy/sdkdb_unittest.py:
(TestSDKDB.test_audit_allowed_conditional):
(TestSDKDB):
(TestSDKDB.test_audit_missing_name_conditional):
(TestSDKDB.test_audit_missing_name_negated_conditional):
(TestSDKDB.test_audit_allowed_negated_conditional):
(TestSDKDB.test_audit_allowed_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions):
(TestSDKDB.test_audit_missing_name_multiple_conditions_negation):

Canonical link: https://commits.webkit.org/297295@main
